lib/watir/locators/element/selector_builder/xpath.rb

```diff
@@ -5,111 +5,113 @@ class SelectorBuilder
 
-            # the remaining entries should be attributes
-            xpath << add_attributes(selector)
+            index = @selector.delete(:index)
 
-            xpath << "[#{index + 1}]" if adjacent && index
+            start_string = default_start
+            adjacent_string = add_adjacent
+            tag_string = add_tag_name
+            class_string = add_class_predicates
+            attribute_string = add_attribute_predicates
+            converted_attribute_string = convert_attribute_predicates
+            text_string = add_text
+            index_string = if !adjacent_string.empty? && index
+                             "[#{index + 1}]"
+                           else
+                             if index
+                               @requires_matches[:index] = index
+                             end
+                             ''
+                           end
 
-            xpath = add_regexp_predicates(xpath, values_to_match)
+            xpath = "#{start_string}#{adjacent_string}#{tag_string}#{class_string}#{attribute_string}" \
+"#{converted_attribute_string}#{text_string}#{index_string}"
 
-            Watir.logger.debug(xpath: xpath, selector: selector)
+            # TODO: See if we can get rid of the merge and just do equals
+            @selector.merge!(@requires_matches)
 
-            [:xpath, xpath]
+            {xpath: xpath}
           end
 
-          def default_start
-            './/*'
-          end
+          # TODO: double check what is actually used in subclasses
+          protected
 
-          def add_tag_name(selector)
-            tag_name = selector.delete(:tag_name).to_s
-            tag_name.empty? ? '' : "[local-name()='#{tag_name}']"
-          end
+          def add_text
+            return '' unless @selector.key?(:text)
 
-          def add_attributes(selector)
-            element_attr_exp = attribute_expression(nil, selector)
-            element_attr_exp.empty? ? '' : "[#{element_attr_exp}]"
+            text = @selector.delete :text
+            if !text.is_a?(Regexp)
+              "[normalize-space()=#{XpathSupport.escape text}]"
+            else
+              @requires_matches[:text] = text
+              ''
+            end
           end
 
-          def add_regexp_predicates(what, selector)
-            return what unless convert_regexp_to_contains?
-
-            selector.each do |key, value|
-              next if %i[tag_name text visible_text visible index].include?(key)
+          def default_start
+            if @selector.key?(:adjacent)
+              './'
+            else
+              './/*'
+            end
+          end
 
-              predicates = regexp_selector_to_predicates(key, value)
-              what = "(#{what})[#{predicates.join(' and ')}]" unless predicates.empty?
+          def add_tag_name
+            tag_name = @selector.delete(:tag_name)
+            if tag_name.is_a?(Regexp)
+              "[contains(local-name(), '#{tag_name.source}')]"
```

Comment:
The any header element is a nice idea. Since you already got it working for simple regexps, I'd just do the quick thing and throw an error if a complex regexp. Presumably you could pull the #simple_regexp? method from the Button XPath class back to here..

Reply:
though if that's really the only use case I think I'd prefer a new like `#section_header` to match just h1-h6.
Really the only time I expect users to use `:tag_name` locator directly is for adjacent calls, which can use optimization if necessary....

I guess the best reason to keep it is to be consistent across our locators. `#simple_regexp?` is in the superclass already, so available to use right there.